pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -338,55 +338,92 @@ func (config *V2RegistriesConf) postProcess() error {
-		return ctx.SystemRegistriesConfDirPath
-	}
-
-	userRegistriesDirPath := filepath.Join(homedir.Get(), userRegistriesDir)
-	if _, err := os.Stat(userRegistriesDirPath); err == nil {
-		return userRegistriesDirPath
-	}
-
-	if ctx != nil && ctx.RootForImplicitAbsolutePaths != "" {
-		return filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfDirPath)
-	}
-
-	return systemRegistriesConfDirPath
+	return newConfigWrapper(ctx).userConfigDirPath
 }
 
 // configWrapper is used to store the paths from ConfigPath and ConfigDirPath
 // and acts as a key to the internal cache.
 type configWrapper struct {
-	configPath    string
+	// path to the registries.conf file
+	configPath string
+	// path to system-wide registries.conf.d directory, or "" if not used
 	configDirPath string
+	// path to user specificed registries.conf.d directory, or "" if not used
+	userConfigDirPath string
 }
 
 // newConfigWrapper returns a configWrapper for the specified SystemContext.
 func newConfigWrapper(ctx *types.SystemContext) configWrapper {
-	return configWrapper{
-		configPath:    ConfigPath(ctx),
-		configDirPath: ConfigDirPath(ctx),
+	var wrapper configWrapper
+	userRegistriesFilePath := filepath.Join(homedir.Get(), userRegistriesFile)
+	userRegistriesDirPath := filepath.Join(homedir.Get(), userRegistriesDir)
+	if ctx == nil {
+		wrapper.configPath = systemRegistriesConfPath
+		wrapper.configDirPath = systemRegistriesConfDirPath
+		if _, err := os.Stat(userRegistriesDirPath); err == nil {
+			wrapper.userConfigDirPath = userRegistriesDirPath
+		}
+		return wrapper
 	}
+
+	if ctx.SystemRegistriesConfPath != "" {
+		wrapper.configPath = ctx.SystemRegistriesConfPath
+		if ctx.SystemRegistriesConfDirPath != "" {
+			wrapper.configDirPath = ctx.SystemRegistriesConfDirPath
+		} else {
+			if ctx.RootForImplicitAbsolutePaths != "" {
+				wrapper.configDirPath = filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfDirPath)
+				wrapper.userConfigDirPath = userRegistriesDirPath
+
+			} else {
+				wrapper.configDirPath = filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfDirPath)
+				wrapper.userConfigDirPath = userRegistriesDirPath
+			}
+		}
+		return wrapper
+	}
+	if _, err := os.Stat(userRegistriesFilePath); err == nil {
+		wrapper.configPath = userRegistriesFilePath
+		if ctx.SystemRegistriesConfDirPath != "" {
+			wrapper.configDirPath = ctx.SystemRegistriesConfDirPath
+		} else {
+			wrapper.userConfigDirPath = userRegistriesDirPath
+		}
+		return wrapper
+	}
+	if ctx.RootForImplicitAbsolutePaths != "" {
+		wrapper.configPath = filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfPath)
+	} else {
+		wrapper.configPath = systemRegistriesConfPath
+	}
+	if ctx.SystemRegistriesConfDirPath != "" {
+		wrapper.configDirPath = ctx.SystemRegistriesConfDirPath
+	} else {
+		if ctx.RootForImplicitAbsolutePaths != "" {
+			wrapper.configDirPath = filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfDirPath)
```

Comment:
[1] E.g. here the code does not set `userConfigDirPath` (I think it should), although the `if ctx.SystemRegistriesConfPath` case does use it. The two should behave consistently (`RootForImplicitAbsolutePaths` does not affect home directories).